bulk_indexer_pool.go

```diff
@@ -0,0 +1,209 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package docappender
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// BulkIndexerPool is a pool of BulkIndexer instances. It is designed to be
+// used in a concurrent environment where multiple goroutines may need to
+// acquire and release indexers.
+//
+// The pool allows a minimum number of BulkIndexers to be guaranteed per ID, a
+// maximum number of indexers per ID and an overall lease limit. This is useful
+// to ensure the pool does not grow too large, even if some IDs are slow to
+// release indexers.
+type BulkIndexerPool struct {
+	indexers chan *BulkIndexer
+	entries  map[string]idEntry
+	mu       sync.RWMutex
+	cond     *sync.Cond   // To wait/signal when a slot is available.
+	leased   atomic.Int64 // Total number of leased indexers across all IDs.
+
+	// Read only fields.
+	min, max, total int64
+	config          BulkIndexerConfig
+}
+
+type idEntry struct {
+	nonEmpty chan *BulkIndexer
+	count    *atomic.Int64
+}
+
+// NewBulkIndexerPool returns a new BulkIndexerPool with:
+// - The specified guaranteed indexers per ID
+// - The maximum number of concurrent BulkIndexers per ID
+// - A total (max) number of indexers to be leased per pool.
+// - The BulkIndexerConfig to use when creating new indexers.
+func NewBulkIndexerPool(guaranteed, max, total int, c BulkIndexerConfig) *BulkIndexerPool {
+	p := BulkIndexerPool{
+		indexers: make(chan *BulkIndexer, total),
+		entries:  make(map[string]idEntry),
+		min:      int64(guaranteed),
+		max:      int64(max),
+		total:    int64(total),
+		config:   c,
+	}
+	p.cond = sync.NewCond(&p.mu)
+	return &p
+}
+
+// Get returns a BulkIndexer for the specified ID. If there aren't any existing
+// indexers for the ID, a new one is created.
+//
+// If the overall limit of indexers has been reached, it will wait until a slot
+// is available. This is a blocking operation.
+func (p *BulkIndexerPool) Get(id string) *BulkIndexer {
```

Comment:
Given that this is a blocking operation, should Get take a context and be cancellable?

Reply:
It is but at the moment is a bit awkward to pass any meaningful context in `appender.runActiveIndexer()`. The previous implementation didn't use any cancellations to receive from the channel and I'm not sure this one should? How do you see this one being used?

Reply:
Currently `runActiveIndexer` uses `active = <-a.available`, and the `a.available` channel is closed by the `Appender.Close` method. IIANM we could get into a state where other busy Appenders may block another one's `runActiveIndexer`. I think one way to address that would be:
 - add a BulkIndexerPool.Register method that creates an entry in `p.entries`
 - change BulkIndexer.Get to no longer create an entry if it does not exist; if it doesn't exist, it should return immediately, and similarly the loop should return early if it finds that the entry has been removed

I'm not 100% sure if it's necessary, just a bit suspicious whenever I see code that blocks but doesn't take a context.

Reply:
good point, I had overlooked that case. Let me have a look and see what I can come up with.